pool.go

```diff
@@ -298,7 +295,101 @@ func (p *pool) fetchWith(ctx context.Context, c cid.Cid, with string) (blk block
+	// wait for pool to be initialised
+	<-p.started
+
+	transientErrs := make(map[string]error)
+
+	left := p.config.MaxRetrievalAttempts
+	aff := with
+	if aff == "" {
+		aff = path
+	}
+
+	p.lk.RLock()
+	if left > len(p.endpoints) {
+		left = len(p.endpoints)
+	}
+
+	// if there are no endpoints in the consistent hashing ring, we submit a pool refresh request and fail this fetch.
+	if p.c == nil || p.c.Size() == 0 {
+		p.lk.RUnlock()
+		return ErrNoBackend
+	}
+	nodes, ok := p.c.GetNodes(aff, left)
+	p.lk.RUnlock()
+
+	// if there are no endpoints in the consistent hashing ring for the given cid, we submit a pool refresh request and fail this fetch.
+	if !ok || len(nodes) == 0 {
+		select {
+		case p.refresh <- struct{}{}:
+		default:
+		}
+
+		return ErrNoBackend
+	}
+
+	carFetchStart := time.Now()
+
+	pq := []string{path}
+	for i := 0; i < len(nodes); i++ {
+		err = p.fetchResourceAndUpdate(ctx, nodes[i], pq[0], i, cb, transientErrs)
+
+		if err == nil {
+			pq = pq[1:]
+			if len(pq) == 0 {
+				durationMs := time.Since(carFetchStart).Milliseconds()
+				// TODO: how to account for total retrieved data
+				//fetchSpeedPerBlockMetric.Observe(float64(float64(len(blk.RawData())) / float64(durationMs)))
+				fetchDurationCarSuccessMetric.Observe(float64(durationMs))
+
+				// downvote all parked failed nodes as some other node was able to give us the required content here.
+				reqs := make([]weightUpdateReq, 0, len(transientErrs))
+				for node, err := range transientErrs {
+					goLogger.Debugw("downvoting node with transient err as fetch was subsequently successful", "node", node, "err", err)
+					reqs = append(reqs, weightUpdateReq{
+						node:    node,
+						failure: true,
+					})
+				}
+
+				p.updateWeightBatched(reqs)
+				return
+			} else {
+				// TODO: potentially worth doing something smarter here based on what the current state
+				// of permanent vs temporary errors is.
+
+				// for now: reset i on partials so we also give them a chance to retry.
+				i = 0
+			}
+		} else if errors.Is(err, ErrPartialResponse{}) {
+			epe, ok := err.(ErrPartialResponse)
+			if !ok {
+				continue
+			}
+			if len(epe.StillNeed) == 0 {
+				// the error was ErrPartial, but no additional needs were specified treat as
+				// any other transient error.
+				continue
+			}
+			pq = pq[1:]
+			pq = append(pq, epe.StillNeed...)
+			// TODO: potentially worth doing something smarter here based on what the current state
+			// of permanent vs temporary errors is.
+
+			// for now: reset i on partials so we also give them a chance to retry.
+			i = 0
+		}
+	}
+
+	fetchDurationCarFailureMetric.Observe(float64(time.Since(carFetchStart).Milliseconds()))
+
+	// Saturn fetch failed after exhausting all retrieval attempts, we can return the error.
+	return
+}
+
+func (p *pool) fetchBlockAndUpdate(ctx context.Context, node string, c cid.Cid, attempt int, transientErrs map[string]error) (blk blocks.Block, err error) {
```

Comment:
Can combine `fetchResourceAndUpdate` and `fetchBlockAndUpdate` into one function by making "doFetch"/"fetchResource" an argument to it.